proposals/p3407.md

```diff
@@ -0,0 +1,199 @@
+        calls.
+
+-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
+
+    -   Requiring that declarations of multiple names use `NS.a` syntax is
+        consistent with the single variable case.
+    -   Requiring namespace members be declared while in the same name scope as
+        the namespace itself makes lifetimes clearer.
+
+## Alternatives considered
+
+### Allow prefixing a tuple binding pattern with a namespace
+
+We could use the namespace to prefix the binding tuple. For example:
+
+```carbon
+var NS.(a: i32, b: i32) = (3, 4);
+```
+
+It's rare that we would have a single statement declare multiple names. As a
+consequence, the separation of the namespace qualifier from the declared
+identifier might end up unique to this syntax. In that context, we prefer `NS.a`
+for consistency with other cases, such as `class NS.class`.
+
+### Allow mixing namespaces within a binding pattern
+
+We could allow mixing namespaces in a binding pattern. For example:
+
+```carbon
+var (e: i32, NS.f: i32) = (5, 6);
+var (FirstNS.g: i32, SecondNS.h: i32) = (7, 8);
+```
+
+That would offer the possibility of mixing use inside scopes:
+
+```carbon
+namespace NS;
+
+class ClassT {
+  var (static class_member: i32,
+       NS.namespace_member: i32,
+       package.package_member: i32) = InitData();
+}
+```
+
+Note here that static syntax for class members is not defined: I'm assuming for
+argument's sake that we expect something similar, but this could also be:
+
+```
+  let (class_member: i32,
+       var NS.namespace_member: i32,
+       var package.package_member: i32) = InitData();
+```
+
+It's not obvious how memory for the above should be allocated. There are two
+options:
+
+-   The three values are sequentially allocated, even though they belong to
+    different name scopes. This allows for return value optimizations to work,
+    but may cause issues for how we consider name scopes and lifetimes.
+-   The three values are separately allocated, grouping allocations by name
+    scope. This breaks return value optimizations, but simplifies requirements
+    around name scopes and lifetimes.
+
+Both options have issues; it would likely prove difficult to set this up such
+that return value optimizations would consistently work, across all potential
+uses of this syntax. That would introduce some variability in performance based
+on unclear context, something we would like to avoid.
+
+Without mixing namespaces, static initialization case can still be addressed,
+albeit with more work. For example:
+
+```carbon
+namespace NS;
+
+let init_data: auto = InitData();
+
+class ClassT {
+  var static class_member: i32 = init_data[0];
+}
+
+var NS.namespace_member: i32 = init_data[1];
+var package_member: i32 = init_data[2];
+```
+
+Rather than forcing a choice, this alternative is discarded in order to require
+syntax -- and resulting performance -- be clearer.
```

Comment:
Could we can have a much simpler/shorter rationale here without fully exploring the static class variable space? Maybe something like:

````suggestion
The mixture of two different prefixes in a single declaration alone seems likely to be confusing, and doesn't have much motivation.

If we allow namespaced names within unrelated scopes, it would also imply the possibility of mixing use inside *class* scopes:

```carbon
namespace NS;

class ClassT {
  var (class_member: i32,
       NS.namespace_member: i32,
       package.package_member: i32);
}
```

Either these are all instance members, which would be surprising in terms of the naming conventions for instance members (for example requiring looking up the namespace name after the object in `x.NS.namespace_member`), or it would require the qualified names to be *static* members (in C++ terminology).

Mixing static and non-static members in a single declaration seems fundamentally problematic due to lifetimes.

We don't have a design for static class members like this yet, and it doesn't seem like we should introduce or motivate them with namespacing within scopes.

For now, we just disallow the mixing namespaces in single declarations for simplicity and to avoid having to resolve these issues. And we disallow namespaced names within unrelated scopes to avoid the name lookup complexity and any potential confusion over C++-style-`static` or other semantics.
````

Maybe could be further simplified or reduced...

Reply:
I think this has actually gone in a different direction than the original text. The example does not have unrelated scopes: `ClassT` is within `package`.

I'm going to need to spend time cleaning this up.